Make confined_target_dirs optional field

MVrachev opened the pull request:
Fixes: https://github.com/theupdateframework/tuf/issues/1155

**Description of the changes being introduced by the pull request**:

The field confined_target_dirs from the MIRROR_SCHEMA  is
a list of strings. Those strings define the accessible target
paths for that mirror. For one target to be available for that mirror,
its path should have as a prefix at least one of the strings defined
in confined_target_dirs.

That's why when confined_target_dirs is a list with one element empty
string (e.g. `['']`) this means all targets files on that mirror are
available and if confined_target_dirs is empty list (e.g. `[]`) this
would be interpreted as none of the target files is available.

This is a confusing API that could easily lead to mistakes.
That's why it's better we promote to not set confined_target_dirs
at all if a user wants targets to be available.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Make confined_target_dirs optional field

The field confined_target_dirs from the MIRROR_SCHEMA  is
a list of strings. Those strings define the accessible target
paths for that mirror. For one target to be available for that mirror,
its path should have as a prefix at least one of the strings defined
in confined_target_dirs.

That's why when confined_target_dirs is a list with one element empty
string (e.g. ['']) this means all targets files on that mirror are
available and if confined_target_dirs is empty list (e.g. []) this
would be interpreted as none of the target files is available.

This is a confusing API that could easily lead to mistakes.
That's why it's better we promote to not set confined_target_dirs
at all if a user wants targets to be available.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

joshuagl reviewed (commented):
Looks good, thanks Martin. It would be good to ensure we're testing that the existing behaviour when `confined_target_dirs: ['']` continues to work. I don't think that's covered by test_mirrors.py any more with this PR?

Commit:
Add test if confined_target_dirs is ['']

Even though we don't want to promote the usage of [''] as a value
for confined_target_dirs, it's good to test against because we
don't want to introduce a breaking change for our users.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I added a commit testing that.

joshuagl commented on tests/test_mirrors.py:
```diff
@@ -60,30 +60,36 @@ def setUp(self):
+    del self.mirrors['mirror4']
     mirror_list = mirrors.get_list_of_mirrors('target', 'a/b', self.mirrors)
     self.assertFalse(mirror_list)
     self.mirrors['mirror1'] = mirror1
```
Do we need to re-add mirror4 here also?

MVrachev replied:
We could in the name of consistency, but we can say we don't need to re-add mirror1 too because the following tests are regarding exception handling and don't actually check the content of `self.mirrors`.

Maybe I can even delete this re-adding because I don't see any usage of it.

joshuagl commented on tests/test_mirrors.py:
```diff
@@ -60,30 +60,36 @@ def setUp(self):
                  'confined_target_dirs' : ['targets/release/',
-                                            'targets/release/']}}
+                                            'targets/release/']},
+     # Make sure we are testing if confined_target_dirs is ['']
```
```suggestion
     # Make sure we are testing when confined_target_dirs is ['']
```

I wonder if this is overloading the current test too much, perhaps it would be cleaner to add a new test case just for the `confined_target_dirs = ['']` scenario?

MVrachev replied:
This test doesn't contain complicated logic it just contains multiple test cases.
If we are going to add a separate test for this case with the new mirror then it makes sense to split the whole current test case into 4 or 5 tests.

joshuagl replied:
Agreed that the logic isn't too complicated. My concern was that by changing the number of expected entries in the mirror, are we making it harder to ensure the results are as we expect? I see now in the unchanged context lines we're asserting not only the number of matched mirrors but also that they are the mirrors we expect. We should extend that check for mirror4.

joshuagl commented on tests/test_mirrors.py:
```diff
@@ -60,30 +60,36 @@ def setUp(self):
     mirror_list = mirrors.get_list_of_mirrors('target', 'a.txt', self.mirrors)
-    self.assertEqual(len(mirror_list), 1)
+    self.assertEqual(len(mirror_list), 2)
     self.assertTrue(self.mirrors['mirror1']['url_prefix']+'/targets/a.txt' in \
                     mirror_list)
```
I think we want/need a similar assertion for mirror4?

joshuagl commented on tests/test_mirrors.py:
```diff
@@ -60,30 +60,36 @@ def setUp(self):
     mirror_list = mirrors.get_list_of_mirrors('target', 'a/b', self.mirrors)
-    self.assertEqual(len(mirror_list), 1)
+    self.assertEqual(len(mirror_list), 2)
     self.assertTrue(self.mirrors['mirror1']['url_prefix']+'/targets/a/b' in \
                     mirror_list)
```
I think we want a similar assertion for mirror4?

joshuagl reviewed (changes requested):
Apologies for not catching this on the first review, but I think we need to check that mirror4 is in the matched mirror list for the [''] cases.

jku reviewed (commented):
with the changes so far discussed, LGTM

MVrachev commented on tests/test_mirrors.py:
```diff
@@ -80,31 +111,13 @@ def test_get_list_of_mirrors(self):
+    del self.mirrors['mirror5']
     mirror_list = mirrors.get_list_of_mirrors('target', 'a/b', self.mirrors)
     self.assertFalse(mirror_list)
-    self.mirrors['mirror1'] = mirror1
```
Deleted this line because the tests below don't actually care about the content of `self.mirrors`.

MVrachev commented on tests/test_mirrors.py:
```diff
@@ -60,16 +59,48 @@ def setUp(self):
                  'targets_path' : 'targets',
                  'confined_target_dirs' : ['targets/release/v2/']},
+     # confined_target_dirs = [] means that none of the targets on
+     # that mirror is available.
```
Decided to leave a comment about this, because I made a mistake during testing connected with how we interpreted `confined_target_dirs = []`.

joshuagl replied:
Thanks for adding the comment. I wonder if we should document this better throughout the code, most notably in updater.py. AFAICT if `confined_target_dirs` is a `None` value the mirror will never be considered/used. Does that match your understanding?

MVrachev replied:
Yes, in these code changes we are checking if `confined_target_dirs` is a falsely value like that:

```
confined_target_dirs = mirror_info.get('confined_target_dirs')
# confined_target_dirs is an optional field
if confined_target_dirs and not in_confined_directory(full_filepath,
          confined_target_dirs):
     ...
```

I don't want us to leave a lot of comments (especially publicly) about what `confined_target_dirs = []` means, because I prefer if we advertise to not use the field if you don't want confinement for your mirrors. 

This unit test is the only place where `confined_target_dirs = []` could be seen and that's why it's worth leaving that comment.

MVrachev commented:
I had to rebase upon the latest develop branch.
I decided to reorder the test cases in `test_mirrors.py` and that's why the lines changed increased.
Also, I added additional test cases for `mirror5` where we were testing about target confinement.

joshuagl reviewed (approved):
Thank you for adding the extra tests!
